### PR TITLE
Add ninjas

### PR DESCRIPTION
**Pull request by ZePedroResende**

### CODE REVIEW

File: app/assets/stylesheets/ninjas.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the ninjas controller here.
```
Comment (frm) [inline app/assets/stylesheets/ninjas.scss:1]: File not needed, delete please.

File: app/assets/stylesheets/scaffolds.scss [hunk 2]
```diff
@@ -0,0 +1,84 @@
+body {
+  background-color: #fff;
```
Comment (frm) [inline app/assets/stylesheets/scaffolds.scss:2]: Where is this coming from?

File: app/controllers/ninjas_controller.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
+class NinjasController < ApplicationController
+  before_action :set_ninja, only: [:show, :edit, :update, :destroy]
+
+  # GET /ninjas
+  # GET /ninjas.json
```
Comment (frm) [inline app/controllers/ninjas_controller.rb:5]: Please remove these comments, the methods are self-explanatory if we are following rest.

File: app/controllers/ninjas_controller.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
+class NinjasController < ApplicationController
+  before_action :set_ninja, only: [:show, :edit, :update, :destroy]
```
Comment (frm) [inline app/controllers/ninjas_controller.rb:2]: Setting resources on callbacks is a bad practice.

Sure, it eliminates duplication, but it also means that you will lose context. If am reading a method that suddenly calls `@ninja`, where is that coming from? And actions like `show` are empty. Naturally, I'll be thinking "ok, this does nothing", yet a variable is used in the view.

I'd rather if you set the variable explicitly in each action.

File: app/controllers/ninjas_controller.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  # POST /ninjas
+  # POST /ninjas.json
+  def create
+    @ninja = Ninja.new(ninja_params)
+
+    respond_to do |format|
+      if @ninja.save
+        format.html { redirect_to @ninja, notice: "Ninja was successfully created." }
+        format.json { render :show, status: :created, location: @ninja }
```
Comment (frm) [inline app/controllers/ninjas_controller.rb:32]: We're not doing `json`, you can just have the `html` part and leave out the `respond_to` block

File: app/controllers/ninjas_controller.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      end
+    end
+  end
+
+  # PATCH/PUT /ninjas/1
+  # PATCH/PUT /ninjas/1.json
+  def update
+    respond_to do |format|
+      if @ninja.update(ninja_params)
+        format.html { redirect_to @ninja, notice: "Ninja was successfully updated." }
```
Comment (frm) [inline app/controllers/ninjas_controller.rb:45]: Same as before

File: app/controllers/ninjas_controller.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      end
+    end
+  end
+
+  # DELETE /ninjas/1
+  # DELETE /ninjas/1.json
+  def destroy
+    @ninja.destroy
+    respond_to do |format|
+      format.html { redirect_to ninjas_url, notice: "Ninja was successfully destroyed." }
```
Comment (frm) [inline app/controllers/ninjas_controller.rb:59]: Same as before

File: app/assets/stylesheets/table.css [hunk 8]
```diff
@@ -0,0 +1,15 @@
+table {
```
Comment (frm) [inline app/assets/stylesheets/table.css:1]: Where's this coming from?

File: app/controllers/ninjas_controller.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      format.json { head :no_content }
+    end
+  end
+
+  def checkin_ninja
```
Comment (frm) [inline app/controllers/ninjas_controller.rb:64]: First, this is out of the scope of this PR. If you are adding a `Ninja` resource, let's do that first and think about check ins later.

Second, it's a typo, it's `check_in`, not `checkin`.

Third, this is not REST. We don't create methods other than the default REST ones. To add something like this, you would create a `CheckInController` that would receive a parameter for the `user_id` and add a `check_in`

File: app/models/ninja.rb [hunk 10]
```diff
@@ -0,0 +1,3 @@
+class Ninja < ApplicationRecord
+  validates :n_ninja, :n_mentor, :nome, :nome_ed, :contacto_ed, presence: true
```
Comment (frm) [inline app/models/ninja.rb:2]: These attribute names shouldn't be this way.

First, I don't know what they are. What is `n_ninja` and `n_mentor`? If I am reading your code, I should be able to understand what is happening and, for the life of me, there's no way I can guess what `n_ninja` is.

Second, we are writing code in English, let's not soil the database with portuguese-named attributes.

File: app/views/layouts/application.html.slim [hunk 11]
```diff
@@ -1,10 +1,9 @@
 doctype html
 html
   head
-    title Bokken
+   title Bokken
```
Comment (frm) [inline app/views/layouts/application.html.slim:4]: Is this indentation right?

File: app/views/ninjas/index.html.slim [hunk 12]
```diff
@@ -0,0 +1,17 @@
+p#notice= notice
+h1 Ninjas
+= link_to 'New Ninja', new_ninja_path
+= stylesheet_link_tag "table"
```
Comment (frm) [inline app/views/ninjas/index.html.slim:4]: Why are we adding stylesheets here?

File: app/views/ninjas/edit.html.slim [hunk 13]
```diff
@@ -0,0 +1,7 @@
+h1 Editing Ninja
+= render 'form', ninja: @ninja
+= link_to 'Show', @ninja
+| \
```
Comment (frm) [inline app/views/ninjas/edit.html.slim:4]: Why this?

File: app/views/ninjas/new.html.slim [hunk 14]
```diff
@@ -0,0 +1,4 @@
+h1 New Ninja
+= render 'form', ninja: @ninja
+= link_to 'Back', ninjas_path
+
```
Comment (frm) [inline app/views/ninjas/new.html.slim:4]: Extra newline

File: app/views/ninjas/show.html.slim [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  = form_for :ninja, url: checkin_mentor_ninja_url(params[:id]), method: :put do |f|
+    div= f.hidden_field :n_mentor
+    div= f.hidden_field :id
+    = f.submit "Check in mentor", class: "btn btn-large btn-primary"
+= link_to 'Edit', edit_ninja_path(@ninja)
+| \
+= link_to 'Destroy', @ninja, method: :delete, data: { confirm: 'Are you sure?' }
+| \
+= link_to 'Back', ninjas_path
+
```
Comment (frm) [inline app/views/ninjas/show.html.slim:35]: Extra newline

File: app/views/ninjas/show.html.slim [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  strong Nome:
+  = @ninja.nome
+p
+  strong Numero de sessoes como ninja:
+  = @ninja.n_ninja
+p
+  strong Numero de sessoes como mentor:
+  = @ninja.n_mentor
+p
+  strong Nome encarregado de educacao:
```
Comment (frm) [inline app/views/ninjas/show.html.slim:12]: Please use proper accentuation.

File: config/initializers/assets.rb [hunk 17]
```diff
@@ -11,4 +11,4 @@
 # Precompile additional assets.
 # application.js, application.css, and all non-JS/CSS in the app/assets
 # folder are already added.
-# Rails.application.config.assets.precompile += %w( admin.js admin.css )
+Rails.application.config.assets.precompile += %w( admin.js admin.css table.css )
```
Comment (frm) [inline config/initializers/assets.rb:14]: You don't need this. Just require the files in the `application.{js,css,scss}` manifests.

File: config/routes.rb [hunk 18]
```diff
@@ -1,3 +1,9 @@
 Rails.application.routes.draw do
+  resources :ninjas do
+	member do
+		put :checkin_ninja
+		put :checkin_mentor
```
Comment (frm) [inline config/routes.rb:5]: These are terrible routes for reasons mentioned before.

**Review by frm — CHANGES_REQUESTED** [review]
This is not quite there. You can see my comments below, but we also should discuss PR title and descriptions.

You expect me to review your code yet you give me no context as to what is happening. The PR title simply says "ninjas".

From reading the code I ASSUME it's about adding the ninja resource. But this is now that the application is in the beginning and it is easy to INFER what is happening. You should always use a descriptive title and if possible a complementing description (although not mandatory if the PR title is self-explanatory. A good example would be if you found a bug and had to explain how and why it happens and how you fixed it).

Finally, you are missing specs and apparently, you added the resource but I see no migration anywhere. Your validations are also a bit weak in the sense that you check for presence but that's it.

**Comment by frm** [comment]
Also, please start doing the PRs from our own repo instead of your fork. You are trying to merge from `ZePedroResende:dev` to `coderdojobraga:master`. To allow us to do further testing you it should be `coderdojobraga:zpr-ninjas` to `coderdojobraga:master`.

Please use the convention `<initials>-<feature>` when naming your branch so we know who's working on what, when using the command line and not GitHub.

**Comment by ZePedroResende** [comment]
Gonna be more carefull with the PR title and description .
You clarified some doubts i had with the handle of the check in , but has you said theres no need to add it in this PR i'll rewrite it later . 
The rest ,now that you point them , are obvious mistakes and i should have known better .
I'll rewrite and correct whats needed and try to submit later today an improve version this time with specs .
Thanks for reviewing my awful code, writing this and sorry for wasting your time

**Comment by frm** [comment]
Mate, you don't need to apologize. Nobody is born with knowledge, you learn by doing and that is fine.

Code reviews are to help you improve your code, not to batter you. It's not a time waste, you're the one implementing it so we should be happy that you do it and help you get it right.

**Comment by frm** [comment]
Closing in favor of #7

**Pull request closed without merging**